builtin/logical/nomad/path_roles.go

```diff
@@ -0,0 +1,160 @@
+package nomad
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathListRoles(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "role/?$",
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathRoleList,
+		},
+	}
+}
+
+func pathRoles(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "role/" + framework.GenericNameRegex("name"),
+		Fields: map[string]*framework.FieldSchema{
+			"name": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Name of the role",
+			},
+
+			"policy": &framework.FieldSchema{
+				Type:        framework.TypeCommaStringSlice,
+				Description: "Policy name as previously created in Nomad. Required",
+			},
+
+			"global": &framework.FieldSchema{
+				Type:        framework.TypeBool,
+				Description: "Policy name as previously created in Nomad. Required",
+			},
+
+			"type": &framework.FieldSchema{
+				Type:    framework.TypeString,
+				Default: "client",
+				Description: `Which type of token to create: 'client'
+or 'management'. If a 'management' token,
+the "policy" parameter is not required.
+Defaults to 'client'.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRolesRead,
+			logical.UpdateOperation: b.pathRolesWrite,
+			logical.DeleteOperation: b.pathRolesDelete,
+		},
+	}
+}
+
+func (b *backend) Role(storage logical.Storage, name string) (*roleConfig, error) {
+	entry, err := storage.Get("role/" + name)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving role: %s", err)
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleConfig
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRolesRead(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	role, err := b.Role(req.Storage, name)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	// Generate the response
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"type":   role.TokenType,
+			"global": role.Global,
+		},
+	}
+	if len(role.Policy) > 0 {
+		resp.Data["policy"] = role.Policy
+	}
+	return resp, nil
+}
+
+func (b *backend) pathRolesWrite(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	tokenType := d.Get("type").(string)
+	name := d.Get("name").(string)
+	global := d.Get("global").(bool)
```

Comment:
`global` is documented as `Required` in the field schema. Is that the case? If yes, can we use `GetOk` to ensure that it is supplied?

Reply:
I think it's better to default it to false on the global tokens, so now is going to get a value one way or the other. Fixed in 3134c7262.